ContactMangerUI [Step1, 2] Jason & Blue

calledBlu opened the pull request:
안녕하세요, 제이크! (@jryoun1)
💰TEAM BUJA💰의 Blue & Jason입니다!
이번 프로젝트는 Step1의 내용이 많지 않아 Step2와 같이 진행하였습니다.
**이번 프로젝트에 대한 커밋은 45981f19f34361af4935999c253eef00c3c4d44d 부터입니다!**
요구사항을 확인했을 때 dummyData를 제공되지 않아 배열로 생성하려다,
최근 학습한 JSON을 활용하여 URL을 받아서 진행하는데 어려움이 있어 파일로 생성하여 적용해보았습니다!
일교차가 심한데 감기 조심하세요!! 리뷰 잘부탁드립니다! 감사합니다!!! 🙇🏻‍♀️ 🙇🏻

### **🧑🏼‍💻 작업 목록**
[ Group Work  ]

🚀 **Step1**
- [x] 이전의 ContactManager Repo를 UI version에 합치기
- [x] ContactManagerUI App Target 추가
- [x] ContactManagerUI Target Membership에 기존 Model 파일 추가

🚀 **Step2**
- [x] TableView를 사용하여 연락처 목록 화면에 표시
    - [x] `UIListContentConfiguration`을 사용하여 Cell을 구성
- [x] cell에 subtitle style 적용
- [x] 각 행(row)에 표시할 항목 출력
    - [x] `JSONSerialization` 을 사용하여 데이터 출력

[Jason]
- [x] `JSONSerialization` 사용법을 학습하여 적용 

[Blue]
- [x] `JSONSerialization -> JSONDecoder`로 Refactoring
- [x] localPath를 Bundle.main.url로 수정

### **💭 학습 키워드**

- Git repository merge
- Target Membership, Target
- UITableViewCell
- Configuration
- JSONDecoder
- JSONSerialization
- Codable (**학습 필요**)

### **🤔고민과 해결**
- 기존에 UILabel 또는 UIImageView처럼 직접 접근하는 방식을 사용할까 아니면 iOS14 version 이상부터 
사용되는 configuration을 사용해야할까 고민이 되었습니다. 하여 AppStore에서 KakoTalk에서 지원하는 버전이 iOS14버전 이상이여서 2020 WWDC Modern Cell Configuration을 시청하고 적용해보았습니다!

- Cell의 속성까지 구현하는데는 문제점이 없었습니다.
dummyData를 제공해주시지 않아서 어떻게 처리를할까 고민이 되었습니다!🤔
위 머릿말에서 먼저 말씀드린것처럼 JSON을 미리 사용해보면서 친숙해보자는 취지로 구현하였습니다. 
그래서 원하지 않았지만... 전역변수로 처리되는 부분도 있고 해당 위치에 있어서 안되고 로직도 분리해야하는것이 맞지만 JSONSerialization을 사용해보며 Codable이 나오기 이전에는 어떻게 Parsing을 하여 LocalData를 불러오는지 경험해보고 셀에 출력되는 부분을 먼저 확인하였습니다!

### **💬 부연 설명**

- Codable 채택 과정에서의 문제점
    > JSON 파일을 파싱할 때 `JSONSerialization`을 사용하여 진행하였는데, Swift4에서 등장한 `Codable`이 `JSONSerialization`에 비해 간단하게 JSON을 파싱할 수 있다고 하여서 해당 방법을 사용하려고 시도해 보았습니다. 시도해 본 코드는 아래와 같습니다.
    ```swift
    class ContactDataSource: NSObject, UITableViewDataSource {
        private var contactInformationList: [ContactInformation] = []

        func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
            return 20
        }

        func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
            let cellIdentifier = "infoCell"
            let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath)

            cell.configurationUpdateHandler = { cell, state in
                var infoContent = cell.defaultContentConfiguration().updated(for: state)
                self.contactInformationList = self.loadJson("Dummy.json")

                infoContent.text = "\(self.contactInformationList[indexPath.row].name)(\(self.contactInformationList[indexPath.row].age))"
                infoContent.secondaryText = "\(self.contactInformationList[indexPath.row].phoneNumber)"

                cell.accessoryType = .disclosureIndicator
                cell.contentConfiguration = infoContent
            }
            return cell
        }

        func loadJson<T: Decodable>(_ filename: String) -> T {
            let data: Data

            guard let file = Bundle.main.url(forResource: filename, withExtension: nil) else {
                fatalError("\(filename) not found.")
            }

            do {
                data = try Data(contentsOf: file)
            } catch {
                fatalError("Could not load \(filename): (error)")
            }

            do {
                return try JSONDecoder().decode(T.self, from: data)
            } catch {
                fatalError("Unable to parse \(filename): (error)")
            }
        }
    }
    ```
    - VC에서 `contactInformationList`를 선언하여 사용하려고 하였으나, `tableView(cellForRowAt:)` 메서드가 `UITableViewDataSource` 프로토콜을 채택하고 있어서 해당 메서드를 수정할 수 없는 문제가 발생하였고, 오류를 해결하기 위하여 메서드 내에서 구조체 배열 형태의 변수를 선언하는 방식으로 사용하였습니다.
`self` 키워드를 많이 사용해야 하므로 효율적인 코드가 아니라고 판단되어 추가 학습 진행 예정입니다.
    -  `tableView(numberOfRowsInSection:)` 메서드에서 인식하는 `contactInformationList` 프로퍼티는 초기화 된 상태여서, 현재 코드와 다르게 배열의 길이를 사용하여 cell의 개수를 지정해 줄 수 없는 문제가 발생하였습니다.
    - `Codable`에 대해 더 학습한 후 해결해 보도록 하겠습니다!

### **🙋🏻 질문 거리**
- 기존의 Storyboard에 있던 ViewController를 삭제하고 TableViewController를 추가하고 TableView를 적용하였는데
SafeArea때문인지 AutoLayout을 설정하지않아도 TableView의 Frame영역?이 잘 적용되어있는 것 같았습니다.
저희가 생각한바로는 기존 ViewController에 TableView를 추가할때는 AutoLayout을 잡아주었는데 아마 TableViewController에서 기본적으로 제공해주는 속성?이 없어서 그렇다고 생각이 되었습니다! 앞으로 AutoLayout 수업이 한 번 더 있어 해결될 것 같지만 질문드립니다!

- 앞선 부연설명 Part에서 아직 해결하지 못한 문제 중, `tableView(numberOfRowsInSection:)` 메서드 내에서 구조체 배열 형태의 변수를 선언하는 부분에 대해 질문드립니다.
`UITableViewDataSource`에서 데이터를 담는 구조체 배열을 선언하는 것보다 VC에서 별도로 객체화한 후 DataSource에서 사용하는 것이 더 적합해 보이는데, VC에서 객체화 한 데이터를 DataSource로 가져오는 방법을 찾아보고 있는데 아직 많이 막연합니다...! 혹시 해당 부분 관련하여 추천해주실 학습 키워드가 있을지 질문드립니다!


### 💻  **실행 화면**
<img src = "https://user-images.githubusercontent.com/92699723/215694483-cbb29a1f-4c57-446c-9eb5-1d11a1b0089c.gif">

tastycodeTeam committed:
Initial commit

tastycodeTeam committed:
chore

calledBlu committed:
Feat: 프로젝트 생성

Commit:
Feat:  user에게 값을 입력받는 기능 구현

네비게이터: Blue, 드라이버: Jason
- 연락처 정보를 받을 구조체
- 프로그램 실행을 담당하는 클래스 밑 함수 구현

calledBlu committed:
Chore: contactInfo 타입을 파일로 분리

네비게이터: Jason, 드라이버: Blue
- 가독성을 위해 파일 분리

calledBlu committed:
Feat: PrintMessage 타입을 통해 출력되는 메세지 구현

- CustomStringConvertible 프로토콜을 채택하여 출력 케이스를 호출하여 사용

calledBlu committed:
Refactor: 기존 main 코드의 기능과 타입을 분리하여 개선하려고 노력

- Identifier 변수를 사용하여 반복문을 제어하는 역할로 사용 -> 추후 변경 예정
- getUserInputValues 메서드를 사용하여 입력만 받는 기능으로 개선
- 기존에 contactManager 메서드명을 가독성을 위해 processStart로 네이밍 변경
- 이외 프로젝트 변경사항에 대하여 같이 커밋

calledBlu committed:
Feat: 입력받은 문자열을 문자 배열로 변환해주는 메서드 구현

- convertToCharacter(this:) 메서드 구현

calledBlu committed:
Style: 변수명 오타 수정

Commit:
Feat: 입력 문자가 담긴 배열에서 공백 제거하는 메서드 구현

- removeBlank(from:) 메서드 구현

calledBlu committed:
Refactor: 앞에 작성한 메서드 함수 및 변수명 개선

- removeBlank -> excludeSpaceWord 메서드로 네이밍 변경

Commit:
Feat: 입력받은 문자 배열에서 구분선의 index값을 찾는 메서드 구현

- findDividerIndexes(of:) 메서드 구현

calledBlu committed:
Refactor: findDividerIndexes 메서드 네이밍 변경

- extractIndexWithSlash로 변경

calledBlu committed:
Feat: Detector 파일을 생성 및 관련 메서드 이동

calledBlu committed:
Feat: Converter 클래스 생성 및 문자열 변환 메서드 추가

Commit:
Style: indexCounter의 위치 변경

Commit:
Feat: 이름에 해당하는 문자 배열을 추출하는 매서드 구현

- extractNameCharacter 메서드 구현

Commit:
Feat: convertor 객체 및 이름값을 추출하는 프로퍼티 추가

- extractedName
- convertedName

calledBlu committed:
Refactor: convertor가 수행해야 할 일이 main에 있어서 코드 이동

calledBlu committed:
Feat: 입력 가능한 프로토콜을 구현 및 포함하고 있는 파일 생성

calledBlu committed:
Feat: extension을 사용하여 protocol 채택

Commit:
Feat: 정규식을 담고있는 열거형 생성 및 정규식 검사 기능 구현

- 각 타입에 맞는 정규식들을 열거형으로 생성
- checkRegularExpression(target: pattern:) 메서드 구현

calledBlu committed:
Feat: Enum의 rawValue를 문자열로 출력하기 위한 프로토콜 채택 및 구현

calledBlu committed:
Refactor: 원하는 index의 단어를 추출할 수 있는 메서드 구현

calledBlu committed:
Feat: index 추출 및 정규식 표현 비교하여 출력까지 확인

Commit:
Style: PR을 보내기 위해 사용되지 않는 주석 삭제 및 공백, 개행 제거

calledBlu committed:
Refactor: 축약어를 사용했던 파일명 및 구조체명을 명시적으로 변경

calledBlu committed:
Refactor: main 파일에 구현되어 있던 ContactManager 클래스를 새로운 파일로 분리하여 main에 실행부만 남김

calledBlu committed:
Refactor: API Design Guideline의 Strive for Fluent Usage에 맞추어 네이밍 변경

calledBlu committed:
Refactor: 기존 out of range 오류 수정 및 joinWords 메서드가 정상적으로 작동하도록 수정

calledBlu committed:
Style: extractIndexWithSlash 메서드 내부의 for 구문 내의 공백 제거

calledBlu committed:
Style: Detector 내의 불필요한 코드 삭제

calledBlu committed:
Refactor: Storage Groupe 내부 Sting Interpolation 제거

calledBlu committed:
Refactor: API Design Guide의 Strive for Fluent Usage에 맞추어 네이밍 변경

Commit:
Refactor: 정규표현식 수정

- 시작(^)과 끝($), 위 표현식만 포함한다(*) 키워드 추가

Commit:
Feat: 요구사항에서 빠져있었던 출력문 추가

- invalidName 케이스는 임의로 작성하여 오류 위치를 좀 더 용이하게 파악하기 위해 작성
- validUserInput 타입메서드는 유저가 입력한 값에 문제가 없는 경우에 출력하기 위해 작성

Commit:
Feat: 정규표현식과 비교하여 올바른 값이 들어오는지 확인하도록 checkCorrectWord 인스턴스 메서드 수정

Commit:
Refactor: 기존 코드 수정

- identifier 프로퍼티를 Bool타입으로 변경하여 좀 더 동작여부를 명확하게 하기 위해 수정
- 아무런 입력이 없을 때 다시 입력을 받을 수 있도록 조건문 추가
- 기존 검증 방법의 변화로 인해 불필요해진 코드 삭제

Commit:
Refactor: 기존 출력문에 개행 추가 및 타입메서드에 조건 추가

Commit:
Style: 개행이 되지 않도록 terminator 옵션 추가

Commit:
Feat: 올바른 입력이 들어왔을 때 지정된 출력 형태로 메세지를 출력해주는 기능 구

Commit:
Refactor: 반환값을 [String]에서 ContactInformation? 타입으로 변경

Commit:
Feat: Step2의 메뉴 출력 및 연락처 추가, 종료 기능 구현

- MenuStart 열거형 구현
- addContact() 메서드로 연락처 추가 기능 구현
- closeProgram() 메서드로 프로그램 종료 기능 구현

Commit:
Feat: .DS_Store gitignore에 추가하였지만 기존 파일 변경 불가로 인한 커밋

Commit:
Feat: 연락처 목록보기 viewContactList(value:) 메서드로 기능 구현

Commit:
Feat: 연락처 목록 형식에 맞게 출력하는 viewContact(list:) 타입 메서드로 구현

Commit:
Feat: 연락처 검색 기능 searchContact(list:) 메서드로 기능 구현

Commit:
Feat: 연락처 검색 기능 메뉴에 추가

calledBlu committed:
Refactor: 잘못된 형식이 입력되는 경우 프로그램이 종료되던 오류 수정

calledBlu committed:
Feat: 존재하지 않는 이름을 검색하는 경우 오류 메세지 출력

- searchContact(list:, word:) 메서드에 guard else 구문을 이용하여 오류 메세지 출력

Commit:
Feat: SystemMenuWorkable 프로토콜을 채택하여 해당 메서드 구현 준수 요청

Commit:
Feat: SystemMenuWorkable 프로토콜 파일로 분리

Commit:
Refactor: 기존에 Array로 작성한 부분을 Set으로 구현

- ContactInformation 구조체에 Hashable 프로토콜 채택

Commit:
Refactor: age 정규표현식 수정

Commit:
Docs: README 초안 작성

Commit:
Chore: .DS_Store 커밋

calledBlu committed:
Docs: README 내용 추가

calledBlu committed:
Build: project repo merge

Commit:
Style: origin ContactManager Repo README pull

Commit:
Feat: DS_Store change non-tracking

Commit:
Style: DS_Store commit

calledBlu committed:
.DS_Store banished!

calledBlu committed:
Build: iOS Target 추가

calledBlu committed:
Style: 불필요한 공백 및 개행, 주석 삭제

Commit:
Style: Project Setting

- Minimum Deployment: iOS 15.0
- Delete UITest Target

Commit:
Style: file grouping

Commit:
Feat: add model group in ContackManagerUI target

calledBlu committed:
Feat: TableView 구현 및 스토리보드와 연결

Commit:
Feat: set TableViewCell style subtitle

Commit:
Refactor: TableViewCell Row 갯수 수정

Commit:
Feat: Cell Identifier 설정

- cell identifier = infoCell

calledBlu committed:
feat: TableViewCell의 text, secondaryText 구성

Commit:
Feat: cell에 configurationUpdateHandler 적용하여 cell을 reuse하여 업데이트할 수 있도록 구현

calledBlu committed:
Feat: cell의 accessoryType으로 disclosureIndicator 적용

calledBlu committed:
Feat: 예시처럼 보이기 위한 JSON dummy data 추가

Commit:
Feat: ContactInformation에 Codadle 채택 후 구현

Commit:
Refactor: JSON 형식 수정

Commit:
Feat: JSON Parsing을 통해 데이터 출력 구현

calledBlu committed:
Refactor: local 위치를 나타내는 pullPath를 프로젝트 내의 파일 위치를 가져오도록 filePath로 수정

Commit:
Style: Naming 수정

calledBlu committed:
Fix: 프로젝트 변경사항 반영

jryoun1 commented on ContactManager/ContactManagerUI/Controller/SceneDelegate.swift:
```diff
@@ -0,0 +1,33 @@
+
+import UIKit
+
+class SceneDelegate: UIResponder, UIWindowSceneDelegate {
```
AppDelegate와 SceneDelegate에서 불필요한 주석을 제거해주셨네요 👏 
그런데 두 파일이 왜 Controller 폴더에 같이 grouping되었는지 궁금합니다.

calledBlu replied:
이전 JuiceMaker 프로젝트를 Fork받았을 때 AppDelegate와 SceneDelegate가 ViewController와 같이 그룹핑되어 있어서 똑같이 진행해보았습니다.

평소 AppDelegate와 SceneDelegate의 명확한 차이점을 모르고있었는데 그래도 이번 기회에 학습하는 기회가 되었습니다. 

[WWDC 2019 Architecting your app multiple windows](https://developer.apple.com/videos/play/wwdc2019/258/?time=8)을 학습하여 iOS 12 이하 버전에서는 하나의 프로세스(App)에 하나의 Window였지만 iOS 13부터 `window`의 개념이 `scene`으로 대체되고 아래의 사진처럼 하나의 앱에서 여러개의 Scene을 가질 수 있게되어 기존에 AppDelegate의 역할 중 UI의 상태를 알 수 있는 UILifeCycle에 대한 부분을 `SceneDelegate` 가 하게 되었다고 알게되었습니다.

같은 내용을 iYeah와 Jenna팀에서 cory 리뷰어분께서 AppConfiguration이라는 폴더에 그룹핑을 하였다고 보게되었습니다. 학습하고나서 저희팀도 위와 비슷하게 구성해도 괜찮을 것 같았고 LaunchScreen.storyboard와 같이 들어가도 될 것 같다는 생각이 들어 딱 정해져있는 방법이 없지만

어떻게하면 좋을지 고민이됩니다!! Jake는 보통 어떻게 사용하실까요?? 🤔

jryoun1 replied:
넵 configuration쪽에 같이 들어가도 될 것 같아보이네요 ㅎㅎ 
새로운 것도 학습해보시고 아주 좋네요 👏💯

jryoun1 commented on ContactManager/ContactManagerUI/Controller/ContactManagerTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+
+import UIKit
+
+var contactInfomation = [String: Any]()
+var nameArray = [String]()
+var ageArray = [String]()
+var phoneNumberArray = [String]()
```
이 변수들은 왜 전역으로 선언되었을까요? 🙋‍♂️

calledBlu replied:
원치 않았지만.. JSONSerialization을 사용해보며 Codable이 나오기 이전에는 어떻게 Parsing을 하여 LocalData를 불러오는지 경험해보고 셀에 출력되는 부분을 먼저 확인해보기위해 전역 변수로 사용하여 확인하게되었습니다.. 

기존의 DataSource에서 제공되는 메서드의 Form(형태)를 변경할 수 없어 어떻게 전달하여 사용하면 좋을까 고민하다보니 DataSource안에 private로 JSONSerialization의 로직을 넣어서 구현한다는게 더 올바르지 않다고 생각되어 로직은 분리하고 Data를 담는 부분만 전역변수로 선언하여 처리하게되었습니다..🥲

jryoun1 commented on ContactManager/ContactManagerUI/Controller/ContactManagerTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+var ageArray = [String]()
+var phoneNumberArray = [String]()
+
+final class ContactManagerTableViewController: UITableViewController {
```
final 키워드를 사용하셨군요 👏
사용하신 이유가 궁금합니다. 
또한 프로젝트 내부에서 어느 객체에는 final이 붙어있고 어느것에는 붙어있지 않는데 사용하시는 기준이 있을까요?

calledBlu replied:
Converter, Detector, Checker 이렇게 final 키워드를 사용하지 않았습니다.
이번 프로젝트에서 로직에 변경이 있을 것 같다고 생각되어 사용하지 않았는데 우선적으로는 접근제어를 걸어놓고 수정사항이 생기면 그때 접근 제한 수준을 낮춰주는 방법을 적용해보겠습니다!

습관을 들이려고해도 짝프로그래밍을 진행하면서 계속 놓치는 부분이라 아쉬운 것 같습니다..😅

jryoun1 replied:
🤔 final을 사용하신 이유는 답변이 아직 안 된 것 같긴합니다! 
final의 경우에는 Swift에서 더이상 해당 클래스는 상속이 되지 않음을 나타내어서, 런타임에 컴파일러가 해당 클래스를 상속받는 클래스가 있는지 없는지를 확인하지 않게 해주기 때문에 성능상의 이점을 얻기 위해서 사용합니다.

위에서 말씀하신 수정사항이라는게 상속일 것 같지는 않은데, 
final이 왜 붙는지에 대해서도 나중에 한 번 공부해보시면 좋을 것 같네요. 🙂

관련내용은 WWDC2016 Understanding Swift performance를 찾아보면 됩니다!

jryoun1 commented on ContactManager/ContactManagerUI/Controller/ContactManagerTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+final class ContactManagerTableViewController: UITableViewController {
+    
+    @IBOutlet private weak var contactManagerTableView: UITableView!
+    private var contactDataSource = ContactDataSource()
```
UITableViewController는 [UITableViewDataSource](https://developer.apple.com/documentation/uikit/uitableviewdatasource)와 
[UITableViewDelegate](https://developer.apple.com/documentation/uikit/uitableviewdelegate)의 프로토콜을 이미 준수하고 있는데, 따로 UITableViewDataSource를 준수하는 DataSource 객체를 생성해서 할당하는 이유가 궁금합니다. 🙋‍♂️

calledBlu replied:
저 TableViewController의 Define을 찾아보았습니다. Delegate와 DataSource 모두 채택하고 있다는 사실을 이번에 처음 알게 되었습니다! 집어주셔서 감사합니다.

공식문서에서도 위 내용을 언급한 부분을 확인하고 숙지하였습니다!

> Subclass `UITableViewController` when your interface consists of a table view and little or no other content. 
> Table view controllers already adopt the protocols you need to manage your table view’s content and respond to changes.

jryoun1 commented on ContactManager/ContactManagerUI/Controller/ContactManagerTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        self.configureTableView()
```
self를 사용하시는 기준이 있을까요?

calledBlu replied:
학습활동 시 예제를 통해 학습하였고, 이번에 구현하면서 해당 코드를 많이 참고하였는데,  self의 용법을 생각하지 못하고 사용했던 것 같습니다!
지금이라도 제대로 알고 사용해야 할 것 같아서 `self.`, `Self`, `.self` 의 사용법에 대해 학습해 보았고, 내용은 아래와 같습니다.
 
- 클래스/구조체 내부에서 객체 자체를 나타내는 `self.`
    - `self`를 통해 init() 메서드에서 초기 값을 할당할 수 있음(메서드의 파라미터 이름과 프로퍼티 이름이 같은 경우 self 키워드가 필요)
    - 타입의 클로저가 있는 상태에서 내부 프로퍼티에 접근하기 위한 키워드 `self.`
    - 열거형의 인스턴스를 각 열거형 케이스와 비교하는 switch문에서의 `self.`
    - 타입 확장을 진행하는 경우의 `self` 는 호출 시 해당 유형에 대한 인스턴스에 대한 작업을 수행
- 타입으로 사용되는 `Self`
    - 현재 속한 타입을 나타내고 싶을 때 사용
- 메타 타입의 값을 얻기 위한 `.self`
    - 해당 타입이 어떤 타입인지를 얻기 위해 사용!
    - `type(of:)`를 사용할 수도 있음

학습 내용 중 해당되는 부분이 없었고, 무의미한 `self.`의 사용이라고 생각되어 제거하였습니다!

jryoun1 replied:
많이 공부해보시고 답변을 해주셔서 제가 다 감사하네요 👏
self의 사용에 대해서 여쭤본 이유는 물론 위의 경우들에 사용할 수도 있지만, 만약 자신만의 기준이 있다면 충분히 사용할 수도 있을 것 같아서 혹시 있는지 궁금해서 여쭤봤습니다. 👍🏻

jryoun1 commented on ContactManager/ContactManagerUI/Controller/ContactManagerTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+        guard let filePath = Bundle.main.url(forResource: "Dummy", withExtension: "json") else { return }
+        
+        if let data = try? Data(contentsOf: filePath) {
+            let json =  try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
```
여기서는 try!와 as!를 사용하시고, 그 위에서는 try?를 사용하셨는데 차이는 무엇일까요?
왜 한곳에서는 try?를 사용하고 다른곳에서는 try!를 사용하셨는지도 궁금합니다.

calledBlu replied:
JSON Selialization 참고사이트를 확인하면서 적용하였는데 강제언래핑을 신경쓰지 못한 것 같습니다..

참고해서 사용하더라도 조심하겠습니다! 집어주셔서 감사합니다! 🙇🏻

jryoun1 commented on ContactManager/ContactManagerUI/Controller/ContactManagerTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cellIdentifier = "infoCell"
+        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath)
```
cellIdentifier는 결국 TableViewCell의 이름과 동일할 것인데 매번 저렇게 직접 작성하게되면, 휴먼에러가 발생할 확률도 있고 어느한쪽이 변경되었을때 다른쪽도 이름을 같이 변경해주어야할 것입니다.

이를 제네릭과 protocol을 사용해서 간편하게 처리하는 방법이 있으니 한 번 참고해보시면 좋을 것 같네요.
[문서1](https://gist.github.com/nil-biribiri/fec9107f1ba039d3a7b29201f06277b9) [문서2](https://hyerios.tistory.com/61)

calledBlu replied:
Jake가 첨부한 문서들을 참고하여 해당 방법으로 진행하려고 하였으나, custom class를 상속받는 커스텀 셀을 만들어서 진행하는 방법들이었습니다.
문서2의 reuseIdentifier는 해당 타입의 셀 이름을 return하는 것으로, 커스텀 셀로 리팩토링하는 경우 문제를 빨리 해결할 수 있으나, 저희의 코드를 수정하지 않고도 해당 문제를 해결할 수 있는지를 알아보고 싶었습니다.

저희의 현재 코드는 스토리보드의 prototype cell을 활용하여 table view를 구성하고 있는 형태여서 문서2의 방법을 참고하되, 커스텀셀을 만들지는 않고 resueIdentifier를 통해 dequeue하는 코드를 작성해 보았습니다.

해당 prototype cell이 UITableViewCell을 상속받게 하고, prototype cell의 identifier를 UITableViewCell로 수정하는 작업을 진행하였고, 매번 cell의 타입을 지정해주는 작업이 불필요하다고 생각되어 메타 타입을 활용하는 방법으로 다시 한번 리팩토링을 진행하였습니다.

메타 타입을 활용하여 개선된 코드는 dequeueReusableCell를 호출할 때 해당 셀의 클래스의 메타타입을 매개변수로 전달하면 셀을 생성할 때 셀의 타입을 지정해 주어야 할 필요가 없으며 알아서 타입추론을 하고, 그에 맞는 reuseIdentifier를 정확하게 가져올 수 있게 되었습니다!

이 부분을 리팩토링하고, 과정을 이해하는 데 많은 시간을 할애했지만 새로운 지식을 얻을 수 있어서 뿌듯하고 재미있던 시간이었습니다 😄 감사해요 Jake!

jryoun1 commented on ContactManager/ContactManagerUI/Controller/ContactManagerTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+        let cellIdentifier = "infoCell"
+        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath)
+        
+        cell.configurationUpdateHandler = { cell, state in
+            var infoContent = cell.defaultContentConfiguration().updated(for: state)
+            infoContent.text = "\(nameArray[indexPath.row])(\(ageArray[indexPath.row]))"
+            infoContent.secondaryText = phoneNumberArray[indexPath.row]
+            
+            cell.accessoryType = .disclosureIndicator
+            cell.contentConfiguration = infoContent
+        }
+        return cell
```
cellConfiguration을 공부해서 사용해보셨군요 👏💯
configurationUpdateHandler를 사용하신 이유가 궁금합니다🙋‍♂️

calledBlu replied:
WWDC에서 Cell을 업데이트하는 방법으로 `updateConfiguration(using state: UICellConfigurationState)` 메서드를 호출해서 사용하는 것을 알게되었습니다.
공식문서를 검색해 본 결과, 해당 메서드를 직접적으로 호출하는 것은 지양하라고 되어있고, 대신 setNeedsUpdateConfiguration()을 사용하라고 되어있었습니다.
이에 따라, `updateConfiguration(using:)`을 재정의 할 수 있고, 값이 설정된 후에 현재 state에 따른 configuration을 업데이트 해주는 `setNeedsUpdateConfiguration()`를 호출해 주는 `configurationUpdateHandler`를 사용하기로 결정하였습니다.

아직 STEP에서 주어진 상호작용에 대한 state가 없어서, 다음 STEP에서 cell에 대한 상호작용이 주어진다는 가정 하에 `configurationUpdateHandler`를 사용하였습니다.

핸들러는 Optional Closure 기반으로 하위 클래스안에 override된 updateConfiguration(using:) 을 대체하여 사용한다고 확인하여 적용하게되었습니다.

```swift
@available(iOS 14.0, tvOS 14.0, *)
extension UITableViewCell {
    @objc(_bridgedConfigurationState) @MainActor dynamic open var configurationState: UICellConfigurationState { get }

    @objc(_bridgedUpdateConfigurationUsingState:) @MainActor dynamic open func updateConfiguration(using state: UICellConfigurationState)

    public typealias ConfigurationUpdateHandler = (_ cell: UITableViewCell, _ state: UICellConfigurationState) -> Void

    /// Optional closure-based alternative to overriding `updateConfiguration(using:)` in a subclass. This handler
    /// is called after `updateConfiguration(using:)`. Setting a new handler triggers `setNeedsUpdateConfiguration`.
    @MainActor public var configurationUpdateHandler: UITableViewCell.ConfigurationUpdateHandler?
}
```

jryoun1 commented on ContactManager/ContactManager/Model/ContactInformation.swift:
```diff
@@ -0,0 +1,20 @@
+
+import Foundation
+
+struct ContactInformation: Hashable, Codable {
```
Codable을 채택하셨군요! 👏 
Codable 프로토콜은 어떤 프로토콜일까요? 언제 사용하면 좋을까요?

calledBlu replied:
Codable은 Swift4부터 추가된 프로토콜로 아래와 같이 정의되었음을 확인했었습니다.

```swift
public typealias Codable =  Decodable& Encodable
```

MVC 구조로 앱을 만들면, 도메인 데이터를 모델 객체로 추상화해서 표현하게 되는데 모델 객체를 JSON 데이터로 인코딩하고, JSON 데이터를 모델 객체로 디코딩할 방법이 필요하고 Swift4 이전에 JSONSelialization을 사용할 때는 적용할 수 없었지만 앞으로 Decodable을 사용하면서 Codable을 채택해서 사용하면 된다고 알고있어 리펙토링 해 볼 예정입니다!

jryoun1 replied:
좋네요! 👏
여기서 질문의 의도는 사실 지금 ContactInformation이 Decodable과 Encodable이 다 필요한 상황인지가 궁금해서 여쭤봤습니다!
둘 중에 하나만 사용하고 있다면 해당 protocol만 채택하는것이 해당 타입을 봤을 때, 
바로 이 타입은 Decoding만 하는지 Encoding만 하는지를 가시적으로 알 수 있어서 말씀드렸습니다 🙏🏻

jryoun1 commented on ContactManager/ContactManager/Model/ContactInformation.swift:
```diff
@@ -0,0 +1,20 @@
+    private(set) var age: String
+    private(set) var phoneNumber: String
+    
+    enum CondingKeys: String, CodingKey {
+        case name = "name"
+        case age = "age"
+        case phoneNumber = "phoneNumber"
+    }
```
CodingKeys도 사용하셨군요👍🏻
CodingKeys는 언제사용하는걸까요?

calledBlu replied:
JSON의 데이터 이름과 Model의 프로퍼티가 매핑이 안되는 상황이 발생하는 것을 경험했었습니다.

하여 CodingKeys를 작성하는 것을 습관화하려는 편입니다!

jryoun1 replied:
여기도 위에서 언급한 것과 비슷한 내용입니다.
지금 상황에서 CodingKeys가 없다고 매핑이 되지 않을까요? 🤔 (CondingKeys로 오타가 있었네요)

위에서 말씀드린 것처럼 만약 key와 model의 property 네이밍이 같다면 해당 코드는 불필요한 코드가 될 것 같아요.

jryoun1 commented on ContactManager/ContactManager/Model/ContactInformation.swift:
```diff
@@ -0,0 +1,20 @@
+//
+//  ContactInfo.swift
```
파일명과 내부의 객체의 명은 통일시켜주세요 🙏🏻

calledBlu replied:
넵! 앞으로 더 꼼꼼하게 확인하도록 하겠습니다! 감사합니다 🙏

jryoun1 reviewed (changes requested):
고생하셨습니다.

step1, step2를 진행하시면서 공부하신 흔적들이 많이 보이네요 👏
다만 몇 가지 개선하면 좋을 부분들이 있어서 코멘트 남깁니다!

Codable과 JSONSerialization도 사용해보시면서 공부해보고 실험해보신것은 좋네요. 
다만 고민하신 부분에서 VC에서 tableView(cellForRowAt:)를 구현하지 못한 부분에 대해서는 밑에 코멘트에도 달아두었으니 확인해보시면 좋을 것 같습니다. (override의 개념도 공부해보시면 좋을것 같네요)

-------
질문거리

>  기존의 Storyboard에 있던 ViewController를 삭제하고 TableViewController를 추가하고 TableView를 적용하였는데
SafeArea때문인지 AutoLayout을 설정하지않아도 TableView의 Frame영역?이 잘 적용되어있는 것 같았습니다.
저희가 생각한바로는 기존 ViewController에 TableView를 추가할때는 AutoLayout을 잡아주었는데 아마 TableViewController에서 기본적으로 제공해주는 속성?이 없어서 그렇다고 생각이 되었습니다! 앞으로 AutoLayout 수업이 한 번 더 있어 해결될 것 같지만 질문드립니다!

다음부터는 질문을 해주실때 궁금한 부분이 무엇이며, 어디까지 어떻게 알아봤는데 어디부분에서 막혀서 궁금한것 같다라고 해주시면 좀 더 답변해드리기 쉬울 것 같네요 🙏🏻
우선 답변을 드리자면 UIViewController와 UITableViewController의 차이에 대해서 공부해보시면 좋을 것 같습니다. 

> 앞선 부연설명 Part에서 아직 해결하지 못한 문제 중, tableView(numberOfRowsInSection:) 메서드 내에서 구조체 배열 형태의 변수를 선언하는 부분에 대해 질문드립니다.
UITableViewDataSource에서 데이터를 담는 구조체 배열을 선언하는 것보다 VC에서 별도로 객체화한 후 DataSource에서 사용하는 것이 더 적합해 보이는데, VC에서 객체화 한 데이터를 DataSource로 가져오는 방법을 찾아보고 있는데 아직 많이 막연합니다...! 혹시 해당 부분 관련하여 추천해주실 학습 키워드가 있을지 질문드립니다!

여기도 마찬가지로 결국 VC에서 tableView(numberOfRowsInSection:) 구현하지 못해서 발생하는 문제인것 같은데. UIViewController와 UITableViewController의 차이점과 이때 UITableViewController를 사용해서 구현하는 방법에 대해서 공부해보시면 좋을 것 같네요.

calledBlu committed:
refactor: UITableViewController를 extension으로 refactoring

- tableView(numberOfRowsInSection: ), tableView(cellForRowAt: ) 메서드를 override
- 코드 리팩토링에 따른 불필요한 프로퍼티 삭제
- configureTableView() 메서드 내의 dataSource 수정

calledBlu committed:
Feat: ReusableTableViewCell 프로토콜 초기구현

calledBlu committed:
Feat: UITableView 확장
- dequeueReusableCell(for: ) 메서드를 커스텀 구현

calledBlu committed:
Feat: UITableViewCell 확장
- UITableViewCell에 ReusableTableViewCell 프로토콜 채택

Commit:
Refactor: cellClass 파라미터 추가

Commit:
Refactor: dequeueReusableCell 메서드 변경사항 적용

Commit:
Refactor: Identifier UITableViewCell로 설정

calledBlu committed:
Refactor: self. 키워드 삭제

Commit:
Chore(config): AppConfiguration 그룹핑

- AppDelegate
- SceneDelegate

Commit:
Merge branch 'main' of https://github.com/calledBlu/ios-contact-manager-ui

Commit:
Refactor: 주석 파일명 일치

Commit:
Refactor: JSONDecoder로 리펙토링

- 전역변수 삭제
- JSONSerialization에서 Decoder로 변경

Commit:
Refactor: Dummy.json 파일 수정

calledBlu committed:
Refactor: private 키워드 추가

calledBlu committed:
Feat: Error 타입 정의

calledBlu committed:
Refactor: Error 핸들링 수정, JSON 데이터 할당하는 메서드 분리

jryoun1 commented on ContactManager/ContactManagerUI/Controller/ContactManagerTableViewController.swift:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    private func assignLoadJSONData() {
```
해당 함수내에서는 지금 할당도 하고있고, 파싱도 하고 있는 것 같습니다.
한 가지 역할만 하도록 개선할 수 있을 것 같아요.

1. 함수명을 바꾸고 반환타입을 추가
2. 함수를 2개로 분리

jryoun1 reviewed (approved):
고생하셨습니다! 
코멘트 드린 것에 대해서도 많은 공부를 해보시고 답변을 해주셔서 제가 다 감사하네요 🙏🏻

잘 수정해주셔서 크게 수정할 것은 없지만, 기존 코멘트에 다시 답글 단 부분과 추가로 남기 코멘트는 다음 pr에서 같이 반영해주셔도 될 것 같아요 👍🏻
그럼 다음 step도 화이팅입니다. 💪🏻
